### PR TITLE
[nfc][InstrProf]Remove 'offsetOf' when parsing indexed profiles

### PR DESCRIPTION
**Pull request by mingmingl-llvm**
- In `Header::readFromBuffer`, read the buffer in the forward direction by using `readNext`.
- When compute the header size, spell out the constant.

With the changes above, we can remove `offsetOf` in InstrProf.cpp

### CODE REVIEW

**Comment by minglotus-6** [comment]
I'm planning to resume https://github.com/llvm/llvm-project/pull/88212 after this clean-up.

 https://github.com/llvm/llvm-project/pull/88212 will add at least 2 new header fields. This simplification can reduce code complexity there.

**Comment by kazutakahirata** [comment]
Thank you for cleaning up these things!  I like where things are going overall.  Now, would you mind creating focused patches that do one thing.

- I think the `GET_VERSION` clean up should call a method in `Header`.  If `Header->getVersion()` is confusing, you might say `Header->getIndexedProfileVersion()` or something.  I wouldn't save the version in `ProfileVersion`.  Instead, I would keep calling `Header->getIndexedProfileVersion()`.
- `read_and_decrement` should be named `readAndDecrment` or `readAndDecrement` according to the LLVM Coding Standards.  Now, I feel a bit uneasy about going backward.  Let's discuss this in a separate patch.

**Comment by minglotus-6** [comment]
> Thank you for cleaning up these things! I like where things are going overall. Now, would you mind creating focused patches that do one thing.

Sure. I created https://github.com/llvm/llvm-project/pull/93613 and https://github.com/llvm/llvm-project/pull/93594.

> 
> * I think the `GET_VERSION` clean up should call a method in `Header`.  If `Header->getVersion()` is confusing, you might say `Header->getIndexedProfileVersion()` or something.  I wouldn't save the version in `ProfileVersion`.  Instead, I would keep calling `Header->getIndexedProfileVersion()`.

https://github.com/llvm/llvm-project/pull/93613 does this.

> * `read_and_decrement` should be named `readAndDecrment` or `readAndDecrement` according to the LLVM Coding Standards.  Now, I feel a bit uneasy about going backward.  Let's discuss this in a separate patch.

I did hesitate between reading forward and backward when write the partial forward compatibility patch (https://github.com/llvm/llvm-project/pull/88212), and chose to retain backward parsing mainly because existing code prefers `switch` with fallthroughs.

https://gcc.godbolt.org/z/qdM4naeM1 roughly demonstrated the IR  for both `switch-with-fallthrough` and a couple of `if` statements. Since this is not performance critical code and the thoughts I hear so far (here, and in a chat room which discusses code patterns/styles) prefer to read forward, I updated the code to use `readNext`.

File: llvm/lib/ProfileData/InstrProf.cpp [hunk 1]
```diff
@@ -1627,19 +1627,12 @@ void OverlapStats::dump(raw_fd_ostream &OS) const {
...
-inline size_t constexpr offsetOf(T1 T2::*Member) {
-  constexpr T2 Object{};
-  return size_t(&(Object.*Member)) - size_t(&Object);
-}
-
 // Read a uint64_t from the specified buffer offset, and swap the bytes in
-// native endianness if necessary.
-static inline uint64_t read(const unsigned char *Buffer, size_t Offset) {
+// native endianness if necessary. Increment the buffer past the read value.
+static inline uint64_t readNext(const unsigned char *&Buffer) {
```
Comment (snehasish) [inline llvm/lib/ProfileData/InstrProf.cpp:1632]: Do we need this wrapper now? Instead we can just use the readNext call directly. You can use a `using` directive to reduce the namespace boilerplate if you want.

File: llvm/lib/ProfileData/InstrProf.cpp [hunk 2]
```diff
@@ -1704,19 +1685,17 @@ size_t Header::size() const {
                   "been added to the header, if not add a case statement to "
                   "fall through to the latest version.");
```
Comment (snehasish) [inline llvm/lib/ProfileData/InstrProf.cpp:1686]: Update the comment and assert message?

File: llvm/lib/ProfileData/InstrProf.cpp (hunk 1 above)
Comment (minglotus-6) [reply llvm/lib/ProfileData/InstrProf.cpp]: Removed this wrapper.

File: llvm/lib/ProfileData/InstrProf.cpp (hunk 2 above)
Comment (minglotus-6) [reply llvm/lib/ProfileData/InstrProf.cpp]: done. I also updated the assert message in `static_assert(std::is_standard_layout_v<Header>`.

File: llvm/lib/ProfileData/InstrProf.cpp [hunk 3]
```diff
@@ -1627,66 +1627,46 @@ void OverlapStats::dump(raw_fd_ostream &OS) const {
...
+      endian::readNext<uint64_t, llvm::endianness::little, unaligned>(Buffer);
+  H.HashOffset =
+      endian::readNext<uint64_t, llvm::endianness::little, unaligned>(Buffer);
+  if (H.getIndexedProfileVersion() >= 8)
+    H.MemProfOffset =
+        endian::readNext<uint64_t, llvm::endianness::little, unaligned>(Buffer);
+  if (H.getIndexedProfileVersion() >= 9)
+    H.BinaryIdOffset =
+        endian::readNext<uint64_t, llvm::endianness::little, unaligned>(Buffer);
+  if (H.getIndexedProfileVersion() >= 10)
```
Comment (snehasish) [inline llvm/lib/ProfileData/InstrProf.cpp:1664]: So version 11 is also handled in this condition right? Maybe add a comment to document.

**Review by snehasish — APPROVED** [review]
lgtm, thanks for improving the code here.

File: llvm/lib/ProfileData/InstrProf.cpp [hunk 4]
```diff
@@ -1627,66 +1627,46 @@ void OverlapStats::dump(raw_fd_ostream &OS) const {
...
   using namespace support;
   static_assert(std::is_standard_layout_v<Header>,
-                "The header should be standard layout type since we use offset "
-                "of fields to read.");
+                "Use standard layout for Header for simplicity");
   Header H;
 
-  H.Magic = read(Buffer, offsetOf(&Header::Magic));
+  H.Magic =
+      endian::readNext<uint64_t, llvm::endianness::little, unaligned>(Buffer);
```
Comment (kazutakahirata) [inline llvm/lib/ProfileData/InstrProf.cpp:1637]: I've made `unaligned` a default parameter a while ago.  May I suggest the following?

```suggestion
      endian::readNext<uint64_t, llvm::endianness::little>(Buffer);
```

The same applies to other calls to `readNext`.

File: llvm/lib/ProfileData/InstrProf.cpp [hunk 5]
```diff
@@ -1627,66 +1627,46 @@ void OverlapStats::dump(raw_fd_ostream &OS) const {
...
-        "Please update the reading code below if a new field has been added, "
-        "if not add a case statement to fall through to the latest version.");
-  case 12ull:
-    H.VTableNamesOffset = read(Buffer, offsetOf(&Header::VTableNamesOffset));
-    [[fallthrough]];
-  case 11ull:
-    [[fallthrough]];
-  case 10ull:
+  static_assert(IndexedInstrProf::ProfVersion::CurrentVersion == Version12,
+                "Please update the reading as needed when a new field is added "
```
Comment (kazutakahirata) [inline llvm/lib/ProfileData/InstrProf.cpp:1650]:
```suggestion
                "Please update the reader as needed when a new field is added "
```

**Review by kazutakahirata — APPROVED** [review]
LGTM modulo minor comments.  Thank you for the clean up!

File: llvm/lib/ProfileData/InstrProf.cpp (hunk 3 above)
Comment (minglotus-6) [reply llvm/lib/ProfileData/InstrProf.cpp:1662]: done.

**Comment by minglotus-6** [comment]
I did a sanity check by invoking `llvm-profdata` to show and merge profiles, using profiles without this PR. Going to merge.

**Pull request merged**